tracer/src/Datadog.Trace/Activity/ActivityListener.cs

```diff
@@ -0,0 +1,294 @@
+using Datadog.Trace.DuckTyping;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.Activity
+{
+    internal static class ActivityListener
+    {
+        private const int InitializationBackoffPerRetry = 10000;
+
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(ActivityListener));
+
+        private static int _initializationRetries = 5;
+
+        private static object? _activityListenerInstance;
+        private static Func<object>? _getCurrentActivity;
+
+        private static int _initialized = 0;
+        private static int _stopped = 0;
+
+        public static bool IsRunning
+        {
+            get
+            {
+                return Interlocked.CompareExchange(ref _initialized, 1, 1) == 1 &&
+                       Interlocked.CompareExchange(ref _stopped, 0, 0) == 0;
+            }
+        }
+
+        internal static object? GetCurrentActivityObject()
+        {
+            try
+            {
+                return _getCurrentActivity?.Invoke();
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex, "Error calling Activity.Current");
+            }
+
+            return null;
+        }
+
+        public static IActivity? GetCurrentActivity()
+        {
+            var activity = GetCurrentActivityObject();
+
+            if (activity is null)
+            {
+                return null;
+            }
+
+            if (activity.TryDuckCast<IActivity6>(out var activity6))
+            {
+                return activity6;
+            }
+
+            if (activity.TryDuckCast<IActivity5>(out var activity5))
+            {
+                return activity5;
+            }
+
+            if (activity.TryDuckCast<IW3CActivity>(out var w3cActivity))
+            {
+                return w3cActivity;
+            }
+
+            return activity.TryDuckCast<IActivity>(out var activity4) ? activity4 : null;
+        }
+
+        public static void Initialize() => Initialize(CancellationToken.None);
+
+        public static void Initialize(CancellationToken cancellationToken)
+        {
+            if (Interlocked.CompareExchange(ref _initialized, 1, 0) == 1)
+            {
+                return;
+            }
+
+            // Try to resolve System.Diagnostics.DiagnosticListener, System.Diagnostics.DiagnosticSource
+            var diagnosticListenerType = Type.GetType("System.Diagnostics.DiagnosticListener, System.Diagnostics.DiagnosticSource", throwOnError: false);
+            if (diagnosticListenerType is null)
+            {
+                // Because we cannot resolve the DiagnosticListener type we allow a later initialization.
+                // For this case we are going to do some retries with a back-off.
+                if (Interlocked.Decrement(ref _initializationRetries) > 0)
+                {
+                    Task.Delay(InitializationBackoffPerRetry, cancellationToken).ContinueWith(
```

Comment:
It's been a long time since I used `ContinueWith()`, but we're always careful about using `await` with `ConfigureAwait(false)`. Do we need to do anything explicitly here to not use a captured synchronization context? Or will this use `TaskScheduler.Default` (thread pool)?

Just making sure we don't miss anything.

Reply:
I think the task scheduler defaults to `TaskScheduler.Current` if we don't specify one.

Reply:
The initialization is called from `Instrumentation.Initialize()` method, so it should be find, we don't expect to have any other scheduler there, just to play safe, I'm forcing the Default one in https://github.com/DataDog/dd-trace-dotnet/pull/2446/commits/0f9aebf26fe77430ca46c17e8da58c90edf3e704